hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/process/ConsumerProcessSupervisor.java

```diff
@@ -76,27 +77,55 @@ public void run() {
         long currentTime = clock.millis();
 
         restartUnhealthy();
+        superviseDyingProcesses();
 
         List<Signal> signalsToProcess = new ArrayList<>();
         taskQueue.drain(signalsToProcess::add);
-        signalsFilter.filterSignals(signalsToProcess, runningProcesses.existingConsumers())
-                .forEach(this::tryToProcessSignal);
+        signalsFilter.filterSignals(signalsToProcess, consumerProcesses.runningSubscriptions(),
+		        consumerProcesses.dyingSubscriptions()).forEach(this::tryToProcessSignal);
 
         logger.debug("Process supervisor loop took {} ms to check all consumers", clock.millis() - currentTime);
     }
 
+    public void shutdown() {
+        consumerProcesses.runningProcessesStream()
+                .forEach(p -> p.accept(Signal.of(STOP, subscriptionName(p))));
+
+        consumerProcesses.dyingProcessesStream()
+                .forEach(p -> p.accept(Signal.of(FORCE_KILL_DYING, subscriptionName(p))));
+
+        executor.shutdown();
+    }
+
+    public List<ManagedSubscriptionStatus> runningSubscriptionsStatus() {
+        return consumerProcesses.listRunningSubscriptionsStatus();
+    }
+
+    public Integer countRunningProcesses() {
+        return consumerProcesses.countRunningSubscriptions();
+    }
+
     private void restartUnhealthy() {
-        runningProcesses.stream()
-                .filter(consumerProcess -> !consumerProcess.isHealthy())
+        consumerProcesses.unhealthyProcessesStream()
                 .forEach(consumerProcess -> {
                     Signal restartUnhealthy =
-                            Signal.of(RESTART_UNHEALTHY, consumerProcess.getSubscription().getQualifiedName());
+                            Signal.of(RESTART_UNHEALTHY, subscriptionName(consumerProcess));
                     logger.info("Lost contact with consumer {}, last seen {}ms ago {}. {}",
                             consumerProcess, consumerProcess.lastSeen(), restartUnhealthy.getLogWithIdAndType());
                     taskQueue.offer(restartUnhealthy);
                 });
     }
 
+    private void superviseDyingProcesses() {
```

Comment:
This method checks if there are processes which where scheduled with KILL/STOP but are still dying and did not died ultimately (e.g. because consumers objects inside processes are waiting for connection to crashed Kafka and never return). The method continuously sends FROCE_KILL_DYING signals for these stale processes.